src/core/selinux-access.c

```diff
@@ -137,31 +137,30 @@ _printf_(2, 3) static int log_callback(int type, const char *fmt, ...) {
 }
 
 static int access_init(sd_bus_error *error) {
-
         if (!mac_selinux_use())
                 return 0;
 
         if (initialized)
                 return 1;
 
         if (avc_open(NULL, 0) != 0) {
-                int saved_errno = errno;
                 bool enforce;
 
-                enforce = security_getenforce() != 0;
-                log_full_errno(enforce ? LOG_ERR : LOG_WARNING, saved_errno, "Failed to open the SELinux AVC: %m");
+                {
+                        PROTECT_ERRNO;
+                        enforce = security_getenforce() != 0;
+                }
+
+                log_full_errno(enforce ? LOG_ERR : LOG_WARNING, errno, "Failed to open the SELinux AVC: %m");
 
-                /* If enforcement isn't on, then let's suppress this
-                 * error, and just don't do any AVC checks. The
-                 * warning we printed is hence all the admin will
-                 * see. */
+                /* If enforcement isn't on, then let's suppress this error, and just don't do any AVC checks.
+                 * The warning we printed is hence all the admin will see. */
                 if (!enforce)
                         return 0;
 
-                /* Return an access denied error, if we couldn't load
-                 * the AVC but enforcing mode was on, or we couldn't
-                 * determine whether it is one. */
-                return sd_bus_error_setf(error, SD_BUS_ERROR_ACCESS_DENIED, "Failed to open the SELinux AVC: %s", strerror_safe(saved_errno));
+                /* Return an access denied error based on the original errno, if we couldn't load the AVC but
+                 * enforcing mode was on, or we couldn't determine whether it is one. */
+                return sd_bus_error_setf(error, SD_BUS_ERROR_ACCESS_DENIED, "Failed to open the SELinux AVC: %m");
```

Comment:
smells as if we should have a flavour of sd_bus_error_setf() that takes *both* an errno and a dbus-style error string, and then makes the best of it.

Reply:
Hmm, we'd need `sd_bus_error_setf_errno`, and `sd_bus_error_setfv_errno` at least. Certainly doable, but let's leave that for another PR. This one is becoming too long already.